template/pkg/init.sls

```diff
@@ -0,0 +1,7 @@
+# -*- coding: utf-8 -*-
+# vim: ft=sls
+
+include:
+  - .install
+  - .config
+  - .service
```

Comment:
`template/pkg/init.sls` shouldn't include `.config` nor `.service`, as no configuration or service management should happen under the `pkg` subdir, right?

Reply:
My method of refactoring on community projects is to make the _smallest set of code changes possible_ while ensuring that there is no change in function.  So this was a direct translation of what `template/init.sls` was doing before this PR.

It is also based on how I've found existing formulas doing things, where installation-config-service are kept close together as a logical block.

However, your point is completely valid and represents the ideal.  A couple of questions before I add to this PR (and to leave time for others to comment):

1. We are still agreeing that running this formula should still perform all three stages, as it did before this PR, right?
1. In order to keep to maintain portability, how do you prefer that the other `sls` files are used in the `include` statements and the `require` requistes?
    1. My preference is _navigating from the root directory_ where `topdir` is resolved as [it is here](https://github.com/myii/template-formula/blob/52d03d895ac60cbcef64ca712881eee56d77c21a/template/map.jinja#L4-L10), for example:
        1. `{{ topdir }}.pkg.install`
        1. `{{ topdir }}.pkg.config`
        1. `{{ topdir }}.pkg.service`
    1. Another method is _relative navigation_, which starts by resolving the next directory up and going from there:
        1. `{{ parentdir }}.pkg.install`
        1. `{{ parentdir }}.pkg.config`
        1. `{{ parentdir }}.pkg.service`
 
That's not very clear.  Let me give examples for both.

Formula root directory-based navigation using `topdir`:

```sls
...
{%- set topdir = tpldir.split('/')[0] %}
{%- from topdir ~ "/map.jinja" import template with context %}
{%- from topdir ~ "/macros.jinja" import files_switch with context %}

include:
  - {{ topdir }}.pkg.install

template-config:
  ...
    - require:
      - sls: {{ topdir }}.pkg.install
```

Parent directory-based navigation using `parentdir` and `parentdot`:

```sls
...
{%- set parentdir = salt['file.dirname'](tpldir) %}
{%- set parentdot = parentdir|replace('/', '.') %}
{%- from parentdir ~ "/map.jinja" import template with context %}
{%- from parentdir ~ "/macros.jinja" import files_switch with context %}

include:
  - {{ parentdot }}.pkg.install

template-config:
  ...
    - require:
      - sls: {{ parentdot }}.pkg.install
```

Reply:
> We are still agreeing that running this formula should still perform all three stages, as it did before this PR, right?

Yes, I think there's no discussion about this point: "the template formula should provide an example of:

* installing/removing a package (app), 
* configuring/deleting the config of such app
* starting/stopping (deleting) the service of such app.

My comment in this PR regarding the structure  is that you included the `config` and `service` states **inside** the `pkg` subdir.

```
            {{ topdir }}.pkg.install
            {{ topdir }}.pkg.config
            {{ topdir }}.pkg.service
```

But from @noelmcloughlin proposal (wich makes sense to me), the structure should be (from the formula's *topdir*)

```
{{ topdir }}.init.sls       # includes {pkg,config,service}.init
{{ topdir }}.clean.sls      # includes {service,config,pkg}.clean

{{ topdir }}.pkg            # `pkg/init.sls` which invokes `pkg.install`
{{ topdir }}.pkg.install
{{ topdir }}.pkg.clean

{{ topdir }}.config         # `config/init.sls` which configs the app
{{ topdir }}.config.clean

{{  topdir }}.service    # `service/init.sls which starts the service
{{topdir }}.service.clean
```

Reply:
@javierbertoli Yes, that's fine.  I pulled @noelmcloughlin into the conversation at the beginning because I knew he had interesting things to offer.

Reply:
I agree with @javierbertoli observations that configuration and service should be separate (from `pkg/ `) but we could capture those review comments for future PR.  Using '{{ topdir }}' instead of relative path has one advantage - in terms of consistency - where foreign states are included in the top states.   

For example the following highlights that `nginx`  and `{{ topdir }}` should co-exist in the top directory (i.e. base).

```
include:
  - nginx.ng
   - {{ topdir }}.pkg.install
```

This relationship *maybe* less obvious when parentdir is used.
```
include:
  - nginx.ng
   - {{ parentdir }}.pkg.install
```
 
However the scenario I am raising is not the common pattern so its minor consideration.

Reply:
@noelmcloughlin Thanks for the feedback.  I think we're reaching some good consensus decisions here.  I don't think it's a problem to add these changes to the PR and will do them shortly.  I'm glad that we're all preferring `topdir` -- it keeps things sane.  I'm seriously considering an upstream PR since this needs to be provided by default.

Reply:
One more thought (per the vision stuff, not this PR) is that salt can support both yaml and json configuration.  All formulas use `yaml/` today but anticipating `json/`  or other data serialisation technology language in the design may make sense - something like this?
```
   template-formula/
        template/
          .. etc ..
             yaml/
                  defaults.yaml
                  osmap.yaml
                  osfamilymap.yaml
            json:
                 defaults.json
                 osmap.json
                 osfamilymap.json
```

Reply:
@noelmcloughlin Yes, @aboe76 kicked of issue #24 for discussions about this.

Reply:
Going back to your discussion with @javierbertoli regarding portability - portability can be maintained by redirecting legacy API (i.e.  old.sls) to an agreed/preferred API (i.e. new.sls) by simple include-

**old.sls**
```

include:
  - .new
```

This enforces depreciation of unwanted API for preferred API. The README must be updated accordingly to ONLY document preferred API - after a considered period of time (i.e. 12 months) the unwanted API can be ejected (i.e.  `cd template; rm old.sls; git add --all; git commit -m"remove depreciated SLS"; git push` )

Reply:
@noelmcloughlin I've linked both of your comments above to two of the open issues in this repo, #21 and #24.

Reply:
Thanks @myii

Reply:
is this what formulas should look like? I'm hesitant about these changes,
@noelmcloughlin I liked the idea of moving the yaml stuff in a directory to keep the formula files clean, 
But having to many subfolders doing one thing makes it harder to implement this template design on the plus ``300``` formulas we already have...

Looking at how other config management tools have fixed this, maybe it is time to incoroporate those best practices:

```ansible```

Has a dedicated folder for defaults and for variables per os/os_family
has a dedicated template folder
has a dedicated file folder 
has a task folder where the main code goes that actually does stuff.

```puppet```

has a dedicated folder for main code (manifests)
has a dedicated folder for files and or templates
has lib folder structure for ruby plugins that can be used directly in the main code
can have hiera data yaml structures for variables.

```chef```
has a dedicated folder for everything attributes,definitions,files,libraries, providers, resources,templates and recipes

```saltstack```
has no dedicated folder for files or templates (arbitrary name)
has no dedicated folder for variables/defaults
has a dedicated folder for the main code where everything lives. (state name)

why because everything is data in saltstack, ans sls file can provide states, but also variables,
a jinja file can be a template but also a macro code block, and yaml files provide only data.
but salt can have json, mako,etc, almost any file type which hold data or structure...

I think it's near impossible to come up with a structured way, only @noelmcloughlin initial idea of having different solution templates, which other people can choose,

having a working model for digesting pillar or external data to a formula is one thing, but structuring the rest how states are build/modeled that highly depends on what the formula manages and to what extend.

Reply:
@aboe76 Are you talking about the directory layout in this comment https://github.com/saltstack-formulas/template-formula/pull/48#pullrequestreview-207182085?

Reply:
I think you're talking about those changes. 
I personally feel that we are going from no best practices to a really complex setup with a lot of default directories. Some are fine, others are somewhat useless for most/simplest formulas.
What am I talking about? 

- the multiple yaml mapfiles (osfamily, osmap...) when in most cases you don't need them
- the yaml and json directories
- even the install/, config/, service/ directories

What I mean is: for most formulas is too much, and maybe a simpler layout could be sufficient for most of them. Though we should have a way to complexify this setup when needed.
I think that this complex layout can 'frighten' a new user.

Reply:
@daks yes I'm talking about https://github.com/saltstack-formulas/template-formula/pull/48#pullrequestreview-207182085 

and the multiple yaml files are only to make the formulas portable across distros without too much hassle

Reply:
I know. Maybe those (map.jinja and all os*.yaml) are the ones which should be in a sub-directory to lightened the layout

Reply:
to ease separation of states and (OS) data

Reply:
@daks thats true, and this is one of the things that needs to be discussed openly and maybe there are more solutions to the same problem.

Reply:
@daks @aboe76 If this needs to be discussed in more detail, it will be far easier to do that in our Slack/IRC/Matrix room than via. GitHub comments...

Reply:
@noelmcloughlin raised a need ('formulas should include 'tear down' capabilities) and with that suggested a solution (the directories structure and states).

While I agree the solution is complex, I OK'ed it because it was *better than what we have now* :smile: 

Another solution for that need is to keep out simple states:
```
{{ topdir }}.init.sls   # includes {pkg,config,service}
{{ topdir }}.package    #  manages pkg install/remove
{{ topdir }}.config     # `manages config/remove of config files
{{ topdir }}.service    # starts/stop/remove the service
```

and use a conditional to manage the desired state, like say, in pseudo for a pillar

```
template:
  pkg:
    ensure: absent  # default: installed
```

the state can be

```
template-package:
  pkg.{{ pkg.ensure }}
```

I used to do that a lot in puppet, and kept things simple. On the directory for the yaml/json files, I'd go with a *single* directory wich can have any of them, with a name like 'local_values' or something meaningful (sad that 'config' is taken :smile:)? and that's it?

My 2 GH cents ;)

Reply:
@aboe76  If there are concerns we should stop and discuss! If nothing else, to capture your concerns because this PR can be superseded by another PR to address your concerns.   Beyond this PR there are many considerations that concern me.  

Here is my take - assume a formula consists of two parts-

**(1) state model**
The standard use case is a single **system** (i.e. `nginx`) with three components-
- system packages/libs,  
- system config
- system daemon

Three subdirectories models the system components and API calls-
```
template-formula/
    template/
           package/     (states:  template.package, template.package.clean)
           config/      (states: template.config, template.config.clean)
           service/     (states: template.service, template.service.clean)
```
**(2) data model**
Data artifacts are consumed by deployment automation model (Salt, but maybe Ansible, etc).  
We could try to figure out what directory structure works best.  
```
template-formula/
    template/
           software/package/     (states:  template.software.package, template.core.package.clean)
           software/config/         (states: template.software.config, template.core.config.clean)
           software/service/        (states: template.software.service, template.core.service.clean)
           model/yaml            
           model/jinja             (*.jinja, *.j2, templates, macros too)
           model/json
           model/mako
```

**Common formula patterns**

This list is not exhaustive.

* Docker runs the daemon
======
A docker container may provide the "service" but depend on `package` and `config` states for binding purposes (see https://docs.docker.com/storage/bind-mounts/). I have seen this pattern with nginx reverse proxy for example.

* Multiple services 
=====
The `iscsi-formula` is the best example I can think of.
```
iscsi-formula/
           isns/package/     
           isns/config/      
           isns/service/ 
           target/package/
           target/config/
           target/service/
           initiator/package/
           initiator/config/
           initiator/service
```
* Toolchains
======
A package state is needed, but the `service` state is not.  Examples are Java, vim, IntelliJ, etc.

* Cross Platform packaging
======
- Tarball 
- docker compose file
- git repo
- snap

I agree that this formula cannot handle everything but this needs discussion over longer period I guess.